File: 3750-closest-equal-element-queries/closest-equal-element-queries.py
class Solution:
    def solveQueries(self, nums: List[int], queries: List[int]) -> List[int]:
        n = len(nums)

        indices = {}
        for i, num in enumerate(nums):
            if num not in indices:
                indices[num] = []
            indices[num].append(i)

        ans = [-1] * n

        for idx in indices.values():
            m = len(idx)

            if m == 1:
                continue
            for i in range(m):

                prev_idx_distance = min(abs(idx[i] - idx[(i - 1 + m) % m]), n - abs(idx[i] - idx[(i - 1 + m) % m]))

                next_idx_distance = min(abs(idx[i] - idx[(i + 1) % m]), n - abs(idx[i] - idx[(i + 1) % m]))

                ans[idx[i]] = min(prev_idx_distance, next_idx_distance)

        return [ans[idx] for idx in queries]

File: 3750-closest-equal-element-queries/closest-equal-element-queries.js
var solveQueries = function (nums, queries) {
    const n = nums.length
    const indices = new Map()
    for (let i = 0; i < n; i++) {
        if (!indices.has(nums[i])) {
            indices.set(nums[i], [])
        }
        indices.get(nums[i]).push(i)
    }
    const ans = new Array(n).fill(-1)

    for (const idx of indices.values()) {
        const m = idx.length
        if (m === 1) continue;

        for (let i = 0; i < m; i++) {
            const curr = idx[i]

            const prev_idx = idx[(i - 1 + m) % m]
            const next_idx = idx[(i + 1) % m]

            let prevIdxDist = Math.abs(curr - prev_idx)
            prevIdxDist = Math.min(prevIdxDist, n - prevIdxDist)

            let nextIdxDist = Math.abs(curr - next_idx)
            nextIdxDist = Math.min(nextIdxDist, n - nextIdxDist)

            ans[curr] = Math.min(prevIdxDist, nextIdxDist)
        }
    }
    return queries.map(idx => ans[idx])
};